simulator/fetch/fetch.cpp

```diff
@@ -71,6 +81,34 @@ void Fetch<ISA>::clock_bp( Cycle cycle)
         bp->update( rp_bp_update->read( cycle));
 }
 
+template <typename ISA>
+Addr Fetch<ISA>::clock_instr_cache( Cycle cycle) 
```

Comment:
Recalling our recent discussion on 'getters' and 'mutators': here is an example of hybridized function. It does two actions:
1. Listens to 'long latency port' and fills tag arrays
2. Checks if there is still a cache miss and controls PC selection

I think it should be separate to two methods:
1. port listener `void Fetch<ISA>::clock_instr_cache( Cycle cycle)`
2. controlling checker

Reply:
Ok, I'll meet your requirements, thank you

Reply:
Pavel Igorevich, I've faced a problem while separating my function: if long latency port is ready, I fill tag arrays, but void Fetch<ISA>::clock_instr_cache( Cycle cycle) doesn't return anything(we read PC from port and that's all). So, I lose my PC in this case. May be the best solution is to do another port, which will hold this PC to the next cycle. Could you give me an advice, please?

Reply:
Basically stall by instruction cache miss is the same thing as stall by data dependency or other things, so you may use the same port (`hold_pc` IIRC)

Reply:
thank you

Reply:
Pavel Igorevich, excuse me for troubling, but I want to clarify one more detail. If there is flush, we'll return flushed_PC, but I have next situation: I get some PC from hold_pc port and have a miss, so I put this PC in long-latency port. After that( in 2 cycles) I have a flush signal, thus I return flushed_PC. The problem is that after cache delay(30 cycles) I get PC from long_latency port and it generates mismatch.
Example:
correct sequence of PC(original): ... 4195264 4195268(flush) 4195268 ...
my incorrect one: ... 4195264 4195268 4195264 4195268
Can you help me?

Reply:
> I get PC from long_latency port and it generates mismatch.

So don't get PC from long_latency port, use it only to fill the cache.
On cache miss, re-try to access cache on the next cycle with the same PC. If there is still a miss, repeat once again. Otherwise, you get a hit and unstall pipeline.

Reply:
All right, thank you. It is very unexpected mistake for me. Everything had been fine before I increased the "cache miss" delay.